sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -95,13 +96,40 @@ object AnalysisContext {
 class Analyzer(
     catalog: SessionCatalog,
     conf: SQLConf,
-    maxIterations: Int)
+    maxIterations: Int,
+    lookupCatalog: Option[(String) => CatalogPlugin] = None)
   extends RuleExecutor[LogicalPlan] with CheckAnalysis {
 
   def this(catalog: SessionCatalog, conf: SQLConf) = {
     this(catalog, conf, conf.optimizerMaxIterations)
   }
 
+  def this(lookupCatalog: Option[(String) => CatalogPlugin], catalog: SessionCatalog,
+      conf: SQLConf) = {
+    this(catalog, conf, conf.optimizerMaxIterations, lookupCatalog)
+  }
+
+  object CatalogRef {
+    def unapply(parts: Seq[String]): Option[(Option[CatalogPlugin], CatalogIdentifier)] =
```

Comment:
Nit: The nesting in the return type is a little tricky to follow, particularly `Option[(Option[...])...]`. Is it worth wrapping `(Option[CatalogPlugin], CatalogIdentifier])` in its own dedicated case class? Then it's `Option[CatalogThing]` (I actually don't know what a good name for such a case class is...) which reads nicer.

Reply:
Same here, for lack of a good name.`CatalogRef` is not ideal either. I will put some thoughts in the next few days. Suggestions are welcome!

Reply:
Fixed